File: scripts/remove_unused_and_rotated_amis.rb
require 'aws-sdk'
require 'awesome_print'

def rotated_images
  ec2 = Aws::EC2::Client.new
  ec2.describe_images(filters: [
    {
      name: 'tag:Rotated',
      values: ['true'],
    },
  ]).images.map()
end

def used_image_ids
  as = Aws::AutoScaling::Client.new
  as.describe_launch_configurations
    .launch_configurations
    .map(&:image_id)
end

def unused_rotated_images
  rotated_images.reject { |i| used_image_ids.include?(i.image_id) }
end

images = unused_rotated_images

puts 'Deregistering amis.'
images.map(&:image_id)
  .map { |id| Aws::EC2::Image.new(id) }
  .tap { |img| ap img }
  .each { |img| img.deregister rescue nil }

puts 'Removing snapshots related with de-registered ami.'
images.map { |i| i.block_device_mappings.map(&:ebs) }
  .flatten
  .compact
  .map(&:snapshot_id)
  .map { |id| Aws::EC2::Snapshot.new(id) }
  .tap { |s| ap s }
  .each { |s| s.delete rescue nil }

File: scripts/remove_unused_and_rotated_amis.js
const AWS = require('aws-sdk');
const _ = require('lodash');

AWS.config.region = process.env.AWS_DEFAULT_REGION;

const EC2 = new AWS.EC2();
const AutoScaling = new AWS.AutoScaling();

function rotatedImages() {
  return EC2.describeImages({
    Filters: [
      {
        Name: 'tag:Rotated',
        Values: ['true'],
      },
    ],
  }).promise()
    .then(data => data.Images);
}

function unusedImageIds() {
  return AutoScaling.describeLaunchConfigurations({}).promise()
    .then(data => {
      return data.LaunchConfigurations.map(lc => lc.ImageId);
    });
}

function deregisterImage(image) {
  return EC2.deregisterImage({
    ImageId: image.ImageId,
  }).promise()
    .then(() => {
      console.log(`deregistered ${image.ImageId}`);
      return image;
    })
    // ignore error and continue
    .catch(() => image)
}

function deleteSnapShot(image) {
  const snapShotId = _.chain(image.BlockDeviceMappings)
    .flatten()
    .map(m => m.Ebs)
    .compact()
    .map(m => m.SnapshotId)
    .first()
    .value();
  return EC2.deleteSnapshot({
    SnapshotId: snapShotId,
  }).promise()
    .then(() => {
      console.log(`deleted ${snapShotId}`);
      return image;
    })
    .catch(() => image)
}

Promise.all([rotatedImages(), unusedImageIds()])
  .then(results => {
    const [rotatedImages, unusedImageIds] = results;
    return rotatedImages.filter(image => {
      return !unusedImageIds.includes(image.ImageId)
    })
  })
  .then(images => {
    const promises = images.map(image => {
      return deregisterImage(image).then(deleteSnapShot)
    });
    return Promise.all(promises);
  })
;
